miscellaneous_scripts/extract_apt_patch_data.py

```diff
@@ -0,0 +1,172 @@
+            f"dpkg -s {package} | grep ^Source", shell=True, capture_output=True, text=True
+        )
+        if source_extraction_cmd_output.returncode in RETURN_CODE_OK:
+            source_package = source_extraction_cmd_output.stdout.strip().split()[1]
+
+        if (
+            any([pckg in impacted_packages for pckg in [package, source_package]])
+            and package in upgradable_packages
+        ):
+            patch_package_list.append(package)
+            ## Add package to the dict that maintains the upgradable package data
+            if package in impacted_packages:
+                if package not in upgradable_packages_data_for_impacted_packages:
+                    upgradable_packages_data_for_impacted_packages[package] = []
+                upgradable_packages_data_for_impacted_packages[package].append(package)
+            ## Add source_package to the dict that maintains the upgradable package data
+            if source_package in impacted_packages:
+                if source_package not in upgradable_packages_data_for_impacted_packages:
+                    upgradable_packages_data_for_impacted_packages[source_package] = []
+                upgradable_packages_data_for_impacted_packages[source_package].append(package)
+    return patch_package_list, upgradable_packages_data_for_impacted_packages
+
+
+def process_generative_mode_type(args):
+    """
+    This method drives the functionality for the generative mode execution. It finds all the installed and the upgradable packages.
+    Thereafter, it sends the relevant data to process_packages method to find out all the impacted packages that can be upgraded.
+    In the ends, it dumps the data at save_result_path location.
+    """
+    impacted_packages = args.impacted_packages
+    assert impacted_packages, "Impacted packages need to be passed for generative mode."
+    installed_packages = get_package_list_using_command(run_command="apt list --installed")
+    upgradable_packages = get_package_list_using_command(run_command="apt list --upgradable")
+
+    upgradable_packages_data_for_impacted_packages = {}
+    patch_package_list = []
+    patch_package_dict = {}
+
+    process_packages(
+        installed_packages,
+        impacted_packages,
+        upgradable_packages,
+        patch_package_list,
+        upgradable_packages_data_for_impacted_packages,
+    )
+
+    patch_package_dict = get_installed_version_for_packages(patch_package_list)
+    for k, _ in patch_package_dict.items():
+        patch_package_dict[k]["previous_version"]=patch_package_dict[k].pop("installed_version")
+
+    apt_patch_details = {
+        "patch_package_dict": patch_package_dict,
+        "upgradable_packages_data_for_impacted_packages": upgradable_packages_data_for_impacted_packages
+    }
+    with open(args.save_result_path, "w") as outfile:
+        json.dump(apt_patch_details, outfile, indent=4)
+
+
+def process_modify_mode_type(args):
```

Comment:
Add a comment about how this does an in place change and also where it is used during building

Reply:
Done